### PR TITLE
[BasicBlockSections] Introduce the basic block sections profile version 1.

### PR DESCRIPTION
**Pull request by rlavaee**
This patch introduces a new version for the basic block sections profile as was requested in D158442, while keeping backward compatibility for the old version.

The new encoding is as follows:
```
m <module_name>
f <function_name_1> <function_name_2>...
c <bb_id_1> <bb_id_2> <bb_id_3>
c <bb_id_4> <bb_id_5>
...
```
Module name specifier (starting with 'm') is optional and allows distinguishing profiles for internal-linkage functions with the same name. If not specified, profile will be applied to any function with the same name.
Function name specifier (starting with 'f') can specify multiple function name aliases.
Finally, basic block clusters are specified by 'c' and specify the cluster of basic blocks, and the internal order in which they must be placed in the same section.

### CODE REVIEW

**Review by shenhanc78 — COMMENTED** [review]
`f <function name 1> <function name 2>`
So func name 1 and func name 2 are aliaes, right?

**Comment by HazardyKnusperkeks** [comment]
There went something wrong, hm?

**Review by asl — DISMISSED** [review]
Consider rebasing your commit before submission.

**Comment by rlavaee** [comment]
> `f <function name 1> <function name 2>` So func name 1 and func name 2 are aliaes, right?

Correct. Fixed the function comment.

**Comment by rlavaee** [comment]
ping. @tmsri

**Comment by rlavaee** [comment]
Another friendly ping for review.

File: llvm/lib/CodeGen/BasicBlockSectionsProfileReader.cpp [hunk 1]
```diff
@@ -46,34 +47,121 @@ BasicBlockSectionsProfileReader::getBBClusterInfoForFunction(
...
+//   c <bb_id_4> <bb_id_5>
+//   ...
+// Module name specifier (starting with 'm') is optional and allows
+// distinguishing profile for internal-linkage functions with the same name. If
+// not specified, it will apply to any function with the same name. Function
+// name specifier (starting with 'f') can specify multiple function name
+// aliases. Basic block clusters are specified by 'c' and specify the cluster of
+// basic blocks, and the internal order in which they must be placed in the same
+// section.
+Error BasicBlockSectionsProfileReader::ReadV1Profile() {
```
Comment (tmsri) [inline llvm/lib/CodeGen/BasicBlockSectionsProfileReader.cpp:64]: We have to put an expiry date on "ReadV0Profile".  create_llvm_prof has to be updated to start producing the new format and we should obsolete and delete the old function soon-ish.

**Review by tmsri — APPROVED** [review]

**Pull request merged**